File: scripts/Python_scripts/seurat_anndata_transformation.py
# For detailed documentation, see: Scanpy2Seurat_transformation.md

from scipy.io import mmwrite, mmread
import scanpy as sc
import pandas as pd
import os
import gzip
import numpy as np
 
def export_anndata(adata, export_path="export", compress_mtx=True):
    os.makedirs(export_path, exist_ok=True)
    
    # store raw counts in layers
    raw_path = f'{export_path}/raw_counts.mtx'
    norm_path = f'{export_path}/norm_expr.mtx'
    if compress_mtx:
        with gzip.open(raw_path + ".gz", "wb") as fh:
            mmwrite(fh, adata.layers['counts'].T)
        with gzip.open(norm_path + ".gz", "wb") as fh:
            mmwrite(fh, adata.X.T)
    else:
        mmwrite(raw_path, adata.layers['counts'].T)
        mmwrite(norm_path, adata.X.T)
    
    pd.Series(adata.var.index).to_csv(f'{export_path}/genes.csv', header=False, index=False)
    pd.Series(adata.obs.index).to_csv(f'{export_path}/cells.csv', header=False, index=False)
    adata.obs.to_csv(f'{export_path}/metadata.csv')
    
    if 'X_umap' in adata.obsm:
        umap_df = pd.DataFrame(
            adata.obsm['X_umap'], 
            columns=['UMAP_1', 'UMAP_2'], 
            index=adata.obs.index
        )
        umap_df.to_csv(f'{export_path}/umap_coordinates.csv')
        print(f"UMAP coordinates saved to {export_path}/umap_coordinates.csv")
    else:
        print("Warning: No UMAP coordinates found in adata.obsm")

    if 'X_tsne' in adata.obsm:
        tsne_df = pd.DataFrame(
            adata.obsm['X_tsne'], 
            columns=['TSNE_1', 'TSNE_2'], 
            index=adata.obs.index
        )
        tsne_df.to_csv(f'{export_path}/tsne_coordinates.csv')
        print(f"T-SNE coordinates saved to {export_path}/tsne_coordinates.csv")
    else:
        print("Warning: No TSNE coordinates found in adata.obsm")
        
    print(f"All files exported to {export_path}")


def _read_mtx(path_base):
    if os.path.exists(path_base + ".gz"):
        with gzip.open(path_base + ".gz", "rb") as fh:
            return mmread(fh)
    return mmread(path_base)


def load_anndata_from_seurat_export(export_path="export"):
    counts_path = os.path.join(export_path, "raw_counts.mtx")
    data_path = os.path.join(export_path, "norm_expr.mtx")

    counts = _read_mtx(counts_path).T.tocsr()
    print(counts.shape)
    data = _read_mtx(data_path).T.tocsr()
    print(data.shape)

    genes = pd.read_csv(os.path.join(export_path, "genes.csv"), header=None)[0].astype(str).to_numpy()
    cells = pd.read_csv(os.path.join(export_path, "cells.csv"), header=None)[0].astype(str).to_numpy()

    obs = pd.read_csv(os.path.join(export_path, "metadata.csv"), index_col=0)
    obs.index = obs.index.astype(str)

    adata = sc.AnnData(X=data, obs=obs)
    adata.var_names = genes
    adata.obs_names = cells
    adata.layers["counts"] = counts

    umap_file = os.path.join(export_path, "umap_coordinates.csv")
    if os.path.exists(umap_file):
        umap_df = pd.read_csv(umap_file, index_col=0)
        adata.obsm["X_umap"] = umap_df.loc[adata.obs_names].to_numpy()

    tsne_file = os.path.join(export_path, "tsne_coordinates.csv")
    if os.path.exists(tsne_file):
        tsne_df = pd.read_csv(tsne_file, index_col=0)
        adata.obsm["X_tsne"] = tsne_df.loc[adata.obs_names].to_numpy()

    return adata

File: scripts/R_scripts/seurat_anndata_transformation.R
# For detailed documentation, see: Scanpy2Seurat_transformation.md

library(Seurat)
library(Matrix)
library(qs)

load_exported_anndata <- function(export_path = "export_anndata", save_path = NULL) {
  # Read raw counts matrix
  raw_path <- paste0(export_path, "/raw_counts.mtx")
  raw_gz <- paste0(raw_path, ".gz")
  raw_counts <- if (file.exists(raw_gz)) {
    readMM(gzfile(raw_gz, "rt"))
  } else {
    readMM(raw_path)
  }
  # Read normalized expression matrix
  norm_path <- paste0(export_path, "/norm_expr.mtx")
  norm_gz <- paste0(norm_path, ".gz")
  norm_expr <- if (file.exists(norm_gz)) {
    readMM(gzfile(norm_gz, "rt"))
  } else {
    readMM(norm_path)
  }

  # Read gene and cell names
  genes <- read.csv(paste0(export_path, "/genes.csv"), header = FALSE, stringsAsFactors = FALSE)$V1
  cells <- read.csv(paste0(export_path, "/cells.csv"), header = FALSE, stringsAsFactors = FALSE)$V1
  
  # Read metadata
  metadata <- read.csv(paste0(export_path, "/metadata.csv"), row.names = 1)
  
  # Transpose raw counts and set row/col names
  rownames(raw_counts) <- genes
  colnames(raw_counts) <- cells
  rownames(norm_expr) <- genes
  colnames(norm_expr) <- cells
  
  # Create Seurat object
  seu <- CreateSeuratObject(counts = raw_counts, meta.data = metadata, assay = "RNA")
  DefaultAssay(seu) <- "RNA"

  counts_mat <- Seurat::GetAssayData(seu, layer = "counts")

  # Seurat may sanitize feature names in counts (e.g. '_' -> '-').
  # Force exact dimnames from counts to prevent feature union on SetAssayData.
  rownames(norm_expr) <- rownames(counts_mat)
  colnames(norm_expr) <- colnames(counts_mat)

  # Set normalized data using SetAssayData on the RNA assay
  seu <- SetAssayData(seu, layer = "data", new.data = norm_expr)

  
  # Load UMAP coordinates if exists
  umap_file <- paste0(export_path, "/umap_coordinates.csv")
  if (file.exists(umap_file)) {
    umap_coords <- read.csv(umap_file, row.names = 1)
    seu@reductions$umap <- Seurat::CreateDimReducObject(
      embeddings = as.matrix(umap_coords),
      key = "UMAP_",
      assay = DefaultAssay(seu)
    )
  }
  
  tsne_file <- paste0(export_path, "/tsne_coordinates.csv")
  if (file.exists(tsne_file)) {
    tsne_coords <- read.csv(tsne_file, row.names = 1)
    seu@reductions$tsne <- Seurat::CreateDimReducObject(
      embeddings = as.matrix(tsne_coords),
      key = "TSNE_",
      assay = DefaultAssay(seu)
    )
  }
  
  # Save with qs if requested
  if (!is.null(save_path)) {
    qs_path <- paste0(save_path, ".qs")
    qsave(seu, qs_path)
    cat("Seurat object saved to", qs_path, "\n")
  }

  return(seu)
}

# Export a Seurat object to files that can be loaded into Scanpy/AnnData
export_seurat <- function(
  seu,
  export_path = "export_seurat",
  assay = "RNA",
  counts_layer = "counts",
  data_layer = "data",
  save_umap = TRUE,
  save_tsne = TRUE,
  coerce_sparse = TRUE,
  compress_mtx = TRUE
) {
  dir.create(export_path, showWarnings = FALSE, recursive = TRUE)

  gzip_file <- function(src, dest) {
    in_con <- file(src, "rb")
    out_con <- gzfile(dest, "wb")
    on.exit({
      try(close(in_con), silent = TRUE)
      try(close(out_con), silent = TRUE)
    }, add = TRUE)
    repeat {
      buf <- readBin(in_con, what = "raw", n = 1e6)
      if (!length(buf)) break
      writeBin(buf, out_con)
    }
  }

  # Helper to support Seurat v4 (slot) and v5 (layer)
  get_assay_matrix <- function(obj, assay_name, layer_name, slot_name) {
    if ("layer" %in% names(formals(Seurat::GetAssayData))) {
      return(Seurat::GetAssayData(obj, assay = assay_name, layer = layer_name))
    }
    return(Seurat::GetAssayData(obj, assay = assay_name, layer = slot_name))
  }

  counts_mat <- get_assay_matrix(seu, assay, counts_layer, "counts")
  data_mat <- get_assay_matrix(seu, assay, data_layer, "data")

  if (coerce_sparse) {
    counts_mat <- Matrix::drop0(as(counts_mat, "CsparseMatrix"))
    data_mat <- Matrix::drop0(as(data_mat, "CsparseMatrix"))
  }

  # Write matrices (features x cells), optionally gzipped
  counts_path <- paste0(export_path, "/raw_counts.mtx")
  data_path <- paste0(export_path, "/norm_expr.mtx")
  if (compress_mtx) {
    tmp_counts <- tempfile(pattern = "raw_counts_", fileext = ".mtx")
    tmp_data <- tempfile(pattern = "norm_expr_", fileext = ".mtx")
    writeMM(counts_mat, tmp_counts)
    writeMM(data_mat, tmp_data)
    gzip_file(tmp_counts, paste0(counts_path, ".gz"))
    gzip_file(tmp_data, paste0(data_path, ".gz"))
    unlink(c(tmp_counts, tmp_data))
  } else {
    writeMM(counts_mat, counts_path)
    writeMM(data_mat, data_path)
  }

  # Genes and cells
  genes <- rownames(counts_mat)
  cells <- colnames(counts_mat)
  write.table(
    genes,
    paste0(export_path, "/genes.csv"),
    row.names = FALSE,
    col.names = FALSE,
    quote = FALSE,
    sep = ","
  )
  write.table(
    cells,
    paste0(export_path, "/cells.csv"),
    row.names = FALSE,
    col.names = FALSE,
    quote = FALSE,
    sep = ","
  )

  # Metadata
  write.csv(seu@meta.data, paste0(export_path, "/metadata.csv"))

  # UMAP coordinates if present
  if (save_umap && "umap" %in% names(seu@reductions)) {
    umap_coords <- Embeddings(seu, reduction = "umap")
    write.csv(umap_coords, paste0(export_path, "/umap_coordinates.csv"))
  }

  # TSNE coordinates if present
  if (save_tsne && "tsne" %in% names(seu@reductions)) {
    tsne_coords <- Embeddings(seu, reduction = "tsne")
    write.csv(tsne_coords, paste0(export_path, "/tsne_coordinates.csv"))
  }

  cat("All files exported to", export_path, "\n")
}
